File: src/core/pricing.ts
import type { PricingRule, UsageSnapshot } from "./types.js";

function globToRegExp(pattern: string): RegExp {
  const escaped = pattern
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    .replaceAll("*", ".*")
    .replaceAll("?", ".");
  return new RegExp(`^${escaped}$`, "i");
}

export function findPricingRule(
  model: string | undefined,
  rules: PricingRule[],
): PricingRule | undefined {
  if (!model) {
    return undefined;
  }
  return rules.find((rule) => globToRegExp(rule.pattern).test(model));
}

export function applyPricing(
  usage: UsageSnapshot,
  rules: PricingRule[],
): UsageSnapshot {
  const model =
    usage.resolvedRef ??
    (usage.provider && usage.model
      ? `${usage.provider}/${usage.model}`
      : usage.model);
  const rule = findPricingRule(model, rules);
  if (!rule) {
    return usage;
  }
  const cost =
    ((usage.inputTokens ?? 0) * rule.inputPerMillion +
      (usage.outputTokens ?? 0) * rule.outputPerMillion +
      (usage.cacheReadTokens ?? 0) * rule.cacheReadPerMillion +
      (usage.cacheWriteTokens ?? 0) * rule.cacheWritePerMillion) /
    1_000_000;
  return {
    ...usage,
    turnCost: cost,
    currency: rule.currency,
  };
}

File: openclaw_hermes_feishu_card/pricing.py
from __future__ import annotations

from fnmatch import fnmatchcase

from .config import PricingRule
from .models import UsageSnapshot


def resolve_pricing_rule(rules: tuple[PricingRule, ...], provider: str, model: str) -> PricingRule | None:
    ref = f"{provider}/{model}".strip("/")
    for rule in rules:
        if fnmatchcase(ref.lower(), rule.pattern.lower()):
            return rule
    return None


def calculate_cost(usage: UsageSnapshot, rule: PricingRule | None) -> float | None:
    if rule is None:
        return None
    return (
        usage.input_tokens * rule.input_per_million
        + usage.output_tokens * rule.output_per_million
        + usage.cache_read_tokens * rule.cache_read_per_million
        + usage.cache_write_tokens * rule.cache_write_per_million
    ) / 1_000_000
